plugin/storage/es/spanstore/service_operation.go

```diff
@@ -0,0 +1,85 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package spanstore
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+
+	jModel "github.com/uber/jaeger/model/json"
+	"github.com/uber/jaeger/pkg/cache"
+	"github.com/uber/jaeger/pkg/es"
+)
+
+// ServiceOperationStorage stores service to operation pairs.
+type ServiceOperationStorage struct {
+	ctx          context.Context
+	client       es.Client
+	logger       *zap.Logger
+	serviceCache cache.Cache
+}
+
+// NewServiceOperationStorage returns a new ServiceOperationStorage.
+func NewServiceOperationStorage(
+	ctx context.Context,
+	client es.Client,
+	logger *zap.Logger,
+	cacheTTL time.Duration,
+) *ServiceOperationStorage {
+	return &ServiceOperationStorage{
+		ctx:    ctx,
+		client: client,
+		logger: logger,
+		serviceCache: cache.NewLRUWithOptions(
```

Comment:
It looks like we initialize this cache with a 12 hour TTL. I'm not sure if this will work as intended. For example, if you write the service name on day 0 at 2300. Not until day 1 at 1100 will you write service::operation for the new index. This means all reads on day 1 will not see any services until 1100. You probably have to use indexName-serviceID as your key in the cache.